### DISCUSSION TITLE
blog/refs-events-and-escape-hatches

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#85] [Announcements]
# blog/refs-events-and-escape-hatches

A technical blog about frontend-development, TypeScript and React

https://tkdodo.eu/blog/refs-events-and-escape-hatches

**Comment by mrclay** [comment]
In the-latest-ref, why shouldn’t I set ref.current immediately after the line creating ref? Why wait until useLayoutEffect?

**Reply by mrclay** [reply]
```js
// truly latest ref pattern
const ref = useRef(val);
ref.current = val;
```
And now ref is always as up to date as it can possibly be.

**Reply by TkDodo** [reply]
Because writing to a ref (and reading from it too, actually) is considered a _side effect_, and render functions cannot have side effects - that's what _effects_ are for.

**Reply by fardad-dev** [reply]
Is Props side effect?
Consider this
```js
const callback = props.callback
```
Is this a side effect?
What about this?
```js
const ref = useRef(callback);
ref.current = callback;
```

**Reply by TkDodo** [reply]
> const callback = props.callback
Is this a side effect?

it's a function. just assigning it to a new `const` is not a side-effect. calling that function during render is a side effect.

> What about this?
const ref = useRef(callback);
ref.current = callback;

writing _to_ a ref during render is considered unsafe. It's mentioned in the [react beta docs](https://beta.reactjs.org/learn/referencing-values-with-refs#differences-between-refs-and-state):

> You shouldn’t read (or write) the current value during rendering.

**Reply by zachsents** [reply]
And why useLayoutEffect in that particular example?

**Reply by TkDodo** [reply]
effects run bottom-up, and the order is layout effects before effects, so if you have parent/child and they both define effect & layout effect, the order would be:

- child layouteffect
- parent layouteffect
- child effect
- parent effect

the latest ref pattern often uses layout effects in case it gets passed around to children so that you can use regular effects in those children and still read the refs. This all falls apart if you try to read in a layout effect in the child though, but it’s a bit better.

Nevertheless, those refs shouldn’t be passed around and `useEffectEvent` is a lot better because it gets the timings right and doesn’t allow passing things around anyways.

**Reply by zachsents** [reply]
you're a wealth of knowledge and your blog rocks. just read a ton of articles. thanks a lot

**Comment by gmoniava** [comment]
Hm, in the section "Photoshop" if I correctly understood, I think the given example isn't good way to show how after mutation you still see latest value, because even if I assign object to `count` for example, such as `var count = {count:0}`,
and update it in immutable way, still inside log I will see latest variable. [Demo](https://codesandbox.io/s/zealous-nova-2he4mr?file=/src/App.js)

**Reply by TkDodo** [reply]
The examples are the same because you update the value `count` themselves, and `count` lives outside of react. Your example creates a new object, but the assignment to `count` is the same as in my example :)

**Reply by gmoniava** [reply]
Maybe I wasn't clear. What was the goal of that example in your code? That when you performed mutation(`count++`), inside `logCount` you still observed the latest value isn't it? 

Quote: "We always see the things from the time they were created, but if they are mutated in place, we would still see the latest value. Let's take the basic example from part 1 and store the value in a mutable variable outside the React component"

Btw I think when updating primitive values it is always considered immutable.

**Reply by TkDodo** [reply]
The goal of the example was to show that if you mutate (=update) a variable, you'll see the latest value (as opposed to the example from part1). It's not so much about immutability but about assigning new values to a variable.

**Reply by gmoniava** [reply]
Ah I didn't read part1. I thought it was trying to say that depending on mutable vs immutable update of `count` you would see different value inside `logCount` - which is not the case on that example (and also AFAIK primitive value updates are always immutable). Anyway, for some reason I read it that way (maybe because it talks about immutability and mutations). You may clarify it a bit more if you like.
